src/pie_utils/processor/document/regex_partitioner.py

```diff
@@ -0,0 +1,165 @@
+from __future__ import annotations
+
+import json
+import logging
+import re
+import statistics
+from typing import Any, Callable, Iterable, Iterator, Match
+
+from pytorch_ie.annotations import LabeledSpan
+
+from pie_utils.statistics import WithStatistics
+
+from ..document import DocumentWithPartition
+
+logger = logging.getLogger(__name__)
+
+
+def get_partitions_with_matcher(
+    document: DocumentWithPartition,
+    matcher: Callable[[str], Iterable[Match]],
+    label_group_id: int | None = None,  # = 1,
+    label_whitelist: list[str] | None = None,
+    skip_initial_partition: bool = False,  # = True
+    initial_partition_label: str = "initial part",
+) -> Iterator[LabeledSpan]:
+    """This method yields LabeledSpans as partitions of the given document. matcher is used to
+    search for a pattern in the document. If pattern is found then it returns Match objects that
+    contains matched groups. The required match group is selected using label_group_id to create a
+    partition label. The span of the partition ranges between two matched pattern. It should be
+    noted that none of the partitions overlap.
```

Comment:
* There are articles missing. 
* Also, logic for setting the label is optional (do we have a test for `label_group_id=None`?) and, hence, should be explained at the end (especially before the main functionality). 
* It is also not correct, that "The span of the partition ranges _between_ two matched pattern".
* regarding "It should be noted that none of the partitions overlap.": This sentence does not make sense to me, as said previously, I dont think we have multiple partitions (but multiple parts). But maybe I am really wrong here. Or do you have ideas for better naming? Renaming partition to partitioning?